### PR TITLE
Add RetryStrategy function

### PR DESCRIPTION
**Pull request by chslink**
add a RetryStrategy interface, allow user custom his own retry strategy

### CODE REVIEW

File: option.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package redislock
+
+import (
+	"time"
+)
+
+const defaultRetryDuration = 100 * time.Millisecond
+
+//RetryStrategy describe the lock retry strategy,100ms is slow for some high performance application
```
Comment (dim) [inline option.go]: sorry, but do you mind fixing all comments and run `go vet ./...`

File: option.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+type RetryStrategy interface {
+	//Next
+	Next() bool
+	//GetRetryBackoff
+	GetRetryBackoff() time.Duration
+}
+
+//NormalRetry every 100ms retry acquire lock
```
Comment (dim) [inline option.go]: can you make this private?

```
type linearRetry struct { backoff time.Duration }

func (r linearRetry) NextBackoff() time.Duration { return r.backoff }

// LinearRetry returns a retry strategy with a linear backoff.
func LinearRetry(backoff time.Duration) RetryStrategy {
	return linearRetry{backoff: backoff}
}

File: option.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
+package redislock
+
+import (
+	"time"
+)
+
+const defaultRetryDuration = 100 * time.Millisecond
+
+//RetryStrategy describe the lock retry strategy,100ms is slow for some high performance application
+type RetryStrategy interface {
```
Comment (dim) [inline option.go]: better:

```
// RetryStrategy allows to customise the lock retry strategy.
type RetryStrategy interface {
  // NextBackoff returns the next backoff duration.
  NextBackoff() time.Duration
}
```

File: option.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+func (r *NormalRetry) GetRetryBackoff() time.Duration {
+	return r.backoff
+}
+
+func NewNormalRetry(backoff time.Duration) RetryStrategy {
+	return &NormalRetry{backoff: backoff}
+}
+
+type LimitRetry struct {
```
Comment (dim) [inline option.go]: same, no need to expose this publicly

File: option.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	}
+	r.count++
+	return r.NormalRetry.Next()
+}
+
+func NewLimitRetry(limit int, backoff time.Duration) RetryStrategy {
+	return &LimitRetry{NormalRetry: &NormalRetry{backoff: backoff}, limitCount: limit}
+}
+
+type StepRetry struct {
```
Comment (dim) [inline option.go]: same, please hide

File: redislock.go [hunk 6]
```diff
@@ -56,12 +56,19 @@ func (c *Client) Obtain(key string, ttl time.Duration, opt *Options) (*Lock, err
...
 
 	value := token + opt.getMetadata()
 	ctx := opt.getContext()
 
+	start := time.Now()
+	usedTime := time.Duration(0)
+
 	var backoff *time.Timer
-	for i, attempts := 0, opt.getRetryCount()+1; i < attempts; i++ {
+	for usedTime < ttl && opt.Next() {
```
Comment (dim) [inline redislock.go]: better:

```
for deadline := time.Now().Add(ttl); time.Now().Before(deadline); {
  backoff := opt.NextBackoff()
  if backoff < 1 {
    break
  }
  ...
}
```
```

**Comment by chslink** [comment]
Thanks again,sorry for did't notice keeping the original code style is very important thing.I will fix these mistakes and add test cases.If there is any comments typo or hard to understand,please help me or rewrite by yourself.It is a bit difficult to write exactly meaning for a non-native English speaker.

**Pull request merged**